In [105]:
#!/usr/bin/env python3

import pandas as pd
import errno    
import os

pd.set_option('display.max_rows', 1000)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

'''
Author: Enrico Ceccolini
    TODO write the description
'''

datadir = "/datasets/eurora_data/db/"
infile_jobs_to_nodes = datadir + "job_nodes.csv"
infile_jobs = datadir + "jobs.csv"

# settings
interval_comment = "April"
cpu_suffix = "0_1min_"

start_time = pd.to_datetime('2014-04-01')
end_time = pd.to_datetime('2014-05-01')


## notice that 43 doesn't exists in the db
# cpus=['01', '02', '03', '04', '05', '06', '07', '08', '09', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '40', '41', '42', '44', '45', '46', '47', '48', '49', '50', '51', '52', '53', '54', '55', '56', '57', '58', '59', '60', '61', '62', '63', '64']
cpus=['01'] # test

In [106]:
# obtain the id of all jobs runned in this node in the past
jobs_to_nodes_whole_data = pd.read_csv(infile_jobs_to_nodes)
print("jobs_to_nodes_whole_data contains {} records".format(jobs_to_nodes_whole_data.shape[0]))

jobs_to_nodes_whole_data contains 469095 records


In [107]:
jobs_to_nodes_whole_data.head(2)

job2node_id  node_id   job_id_string  ncpus  ngpus  nmics  mem_requested
0            1       17  498458.node129     16      0      0       14680064
1            2        9  498459.node129     16      0      0       14680064

In [147]:
jobs_to_nodes_whole_data[jobs_to_nodes_whole_data['node_id'] == 129]

job2node_id  node_id    job_id_string  ncpus  ngpus  nmics  mem_requested
205653       205654      129   935758.node129      1      0      0        2097152
205853       205854      129   935931.node129      1      0      0        2097152
205855       205856      129   935933.node129      1      0      0        2097152
206363       206364      129   936221.node129      1      0      0        2097152
207615       207616      129   936616.node129      1      0      0        2097152
207874       207875      129   936828.node129      1      0      0        2097152
208459       208460      129   937418.node129      1      0      0        2097152
208707       208708      129   937609.node129      1      0      0        2097152
213289       213290      129   942184.node129      1      0      0        2097152
221919       221920      129   955125.node129      1      0      0        2097152
226003       226004      129  1007580.node129      1      0      0        2097152
226655       226656      129  1007852.node129      1      0      0        2097152
226928       226929      129  1008047.node129      1      0      0        2097152
227218       227219      129  1008311.node129      1      0      0        2097152
227859       227860      129  1008515.node129      1      0      0        2097152
228025       228026      129  1008637.node129      1      0      0        2097152
228870       228871      129  1008927.node129      1      0      0        2097152
229845       229846      129  1009745.node129      1      0      0        2097152
234798       234799      129  1014440.node129      1      0      0        2097152
235287       235288      129  1014882.node129      1      0      0        2097152
236836       236837      129  1016406.node129      1      0      0        2097152
237962       237963      129  1016771.node129      1      0      0        2097152
242350       242351      129  1020229.node129      1      0      0        2097152
250589       250590      129  1028291.node129      1      0      0        2097152
259649       259650      129  1200563.node129      1      0      0        2097152
259815       259816      129  1200594.node129      1      0      0        2097152
260434       260435      129  1200896.node129      1      0      0        2097152
261586       261587      129  1201177.node129      1      0      0        2097152
271022       271023      129  1209823.node129      1      0      0        2097152
271802       271803      129  1210386.node129      1      0      0        2097152
272066       272067      129  1210509.node129      1      0      0        2097152
272597       272598      129  1210736.node129      1      0      0        2097152
272963       272964      129  1210940.node129      1      0      0        2097152
346534       346535      129  1296233.node129      1      0      0        2097152
346681       346682      129  1296314.node129      1      0      0        2097152
347028       347029      129  1296555.node129      1      0      0        2097152
347249       347250      129  1296731.node129      1      0      0        2097152
347542       347543      129  1296949.node129      1      0      0        2097152
347785       347786      129  1297085.node129      1      0      0        2097152
357230       357231      129  1318347.node129      1      0      0        2097152
368126       368127      129  1330937.node129      1      0      0        2097152
370116       370117      129  1332877.node129      1      0      0        2097152
376439       376440      129  1339201.node129      1      0      0        2097152
376620       376621      129  1339321.node129      1      0      0        2097152
376842       376843      129  1339465.node129      1      0      0        2097152
377085       377086      129  1339685.node129      1      0      0        2097152
438017       438018      129  1628259.node129      1      0      0        2097152
438026       438027      129  1628259.node129      1      0      0      

In [161]:
jobs_to_nodes_whole_data[jobs_to_nodes_whole_data['job_id_string'] == ""]

Empty DataFrame
Columns: [job2node_id, node_id, job_id_string, ncpus, ngpus, nmics, mem_requested]
Index: []

Nodo 129???

In [165]:
# jobs_to_nodes_whole_data.sort_values('node_id')

job2node_id  node_id    job_id_string  ncpus  ngpus  nmics  mem_requested
72818         72819        1   601038.node129     12      0      0        1048576
57260         57261        1   584228.node129      4      0      0        4194304
57261         57262        1   584229.node129      4      0      0        4194304
57262         57263        1   584230.node129      4      0      0        4194304
93818         93819        1   624035.node129     16      0      0        1048576
93808         93809        1   624025.node129      8      0      0        5242880
350875       350876        1  1310993.node129      4      0      1       14680064
57513         57514        1   584873.node129     16      0      0       14680064
350906       350907        1  1311007.node129      4      0      1       14680064
57541         57542        1   584900.node129      4      0      0        4194304
57542         57543        1   584901.node129      4      0      0        4194304
57543         57544        1   584902.node129      4      0      0        4194304
350930       350931        1  1311397.node129      8      0      1        8388608
350931       350932        1  1311399.node129      8      0      1        8388608
57570         57571        1   584929.node129      4      0      0        4194304
57571         57572        1   584930.node129      4      0      0        4194304
350724       350725        1  1310856.node129     16      0      1       14680064
350636       350637        1  1310782.node129      8      0      0        7340032
350635       350636        1  1310781.node129     16      0      1       14680064
94196         94197        1   624562.node129      8      0      0        5242880
349536       349537        1  1309579.node129     16      0      0       10485760
349559       349560        1  1309605.node129     16      0      1       14680064
349563       349564        1  1309612.node129      4      0      1       14680064
349571       349572        1  1309621.node129     16      0      1       14680064
349591       349592        1  1309643.node129     16      0      1       14680064
349607       349608        1  1309662.node129     16      0      0       12582912
349613       349614        1  1309669.node129     16      0      1       14680064
57572         57573        1   584931.node129      4      0      0        4194304
349614       349615        1  1309670.node129     16      0      1       14680064
349627       349628        1  1309683.node129     16      0      1       14680064
349668       349669        1  1309723.node129     16      0      1       14680064
95356         95357        1   626788.node129      2      0      0        4194304
349810       349811        1  1309914.node129     16      0      1       14680064
349975       349976        1  1310087.node129     16      0      1       14680064
349992       349993        1  1310104.node129      4      0      1       14680064
350371       350372        1  1310511.node129      4      0      1       14680064
349615       349616        1  1309671.node129     16      0      1       14680064
349514       349515        1  1309548.node129     16      0      0       10485760
389118       389119        1  1542033.node129      8      0      0        7340032
350969       350970        1  1311428.node129     16      0      0        1048576
205720       205721        1   935800.node129     16      0      0       14680064
58154         58155        1   585981.node129     16      0      0        5242880
58259         58260        1   586074.node129     16      0      0        5242880
58260         58261        1   586075.node129     16      0      1        5242880
92712         92713        1   622696.node129     12      0      0       14680064
92708         92709        1   622691.node129     12      0      0       14680064
92705         92706        1   622524.node129     12      0      0       14680064
92610         92611        1   622337.node129     12      0      0      

In [173]:
df = jobs_to_nodes_whole_data.groupby(['job_id_string','node_id']).size().reset_index(name='counts')
df = df.sort_values('counts')
df[df['node_id'] == 1]

job_id_string  node_id  counts
295188   597120.node129        1       1
295185   597118.node129        1       1
294994   596891.node129        1       1
294992   596889.node129        1       1
294370   596058.node129        1       1
296588   598349.node129        1       1
296571   598332.node129        1       1
295859   597686.node129        1       1
295780   597646.node129        1       1
295778   597644.node129        1       1
295810   597658.node129        1       1
295808   597656.node129        1       1
295807   597655.node129        1       1
296363   598168.node129        1       1
296278   598041.node129        1       1
296259   598019.node129        1       1
296289   598049.node129        1       1
296287   598047.node129        1       1
296285   598045.node129        1       1
291484   592769.node129        1       1
291490   592776.node129        1       1
291491   592777.node129        1       1
292050   593352.node129        1       1
292049   593351.node129        1       1
291960   593267.node129        1       1
291959   593266.node129        1       1
292003   593309.node129        1       1
292000   593307.node129        1       1
292063   593365.node129        1       1
292062   593364.node129        1       1
292087   593372.node129        1       1
291756   593044.node129        1       1
291910   593221.node129        1       1
291885   593205.node129        1       1
291877   593198.node129        1       1
291875   593197.node129        1       1
291873   593196.node129        1       1
290708   591786.node129        1       1
290675   591758.node129        1       1
290672   591754.node129        1       1
290681   591765.node129        1       1
290518   591424.node129        1       1
290515   591420.node129        1       1
290510   591415.node129        1       1
290545   591451.node129        1       1
290546   591452.node129        1       1
291145   592385.node129        1       1
291143   592383.node129        1       1
291104   592359.node129        1       1
291095   592352.node129        1       1
291136   592376.node129        1       1
291120   592368.node129        1       1
291290   592523.node129        1       1
291219   592434.node129        1       1
291215   592430.node129        1       1
291206   592425.node129        1       1
291223   592438.node129        1       1
291234   592453.node129        1       1
291227   592446.node129        1       1
290894   592117.node129        1       1
290880   592103.node129        1       1
290896   592119.node129        1       1
290913   592141.node129        1       1
290912   592140.node129        1       1
291047   592290.node129        1       1
291034   592266.node129        1       1
291065   592312.node129        1       1
290992   592223.node129        1       1
291032   592263.node129        1       1
293286   594899.node129        1       1
293471   595025.node129        1       1
293423   595005.node129        1       1
293124   594821.node129        1       1
293250   594883.node129        1       1
293241   594871.node129        1       1
293263   594886.node129        1       1
292898   594625.node129        1       1
303153   606156.node129        1       1
301714   603653.node129        1       1
301712   603651.node129        1       1
301711   603650.node129        1       1
301820   603758.node129        1       1
301819   603757.node129        1       1
301818   603756.node129        1       1
301817   603755.node129        1       1
301816   603754.node129        1       1
301813   603752.node129        1       1
301812   603751.node129        1       1
304743   608227.node129        1       1
304742   608226.node129        1       1
297722   599676.node129        1       1
297767   599747.node129        1       1
297571   599541.node129        1       1
297526   599502.node129        1       1
297540   599510.node129        1       1
298111   600123.node129        1       1
298044   600045.node129        1 

In [175]:
jobs_to_nodes_whole_data[jobs_to_nodes_whole_data['job_id_string'] == '1008449.node129']

job2node_id  node_id    job_id_string  ncpus  ngpus  nmics  mem_requested
227636       227637       33  1008449.node129      1      0      0        2097152
227637       227638       33  1008449.node129      1      0      0        2097152
227638       227639       33  1008449.node129      1      0      0        2097152
227639       227640       33  1008449.node129      1      0      0        2097152
227640       227641       33  1008449.node129      1      0      0        2097152
227641       227642       33  1008449.node129      1      0      0        2097152
227642       227643       33  1008449.node129      1      0      0        2097152
227643       227644       34  1008449.node129      1      0      0        2097152
227644       227645       34  1008449.node129      1      0      0        2097152
227645       227646       34  1008449.node129      1      0      0        2097152
227646       227647       34  1008449.node129      1      0      0        2097152
227647       227648       34  1008449.node129      1      0      0        2097152
227648       227649       34  1008449.node129      1      0      0        2097152
227649       227650       34  1008449.node129      1      0      0        2097152
227650       227651        1  1008449.node129      1      0      0        2097152
227651       227652        1  1008449.node129      1      0      0        2097152
227652       227653        1  1008449.node129      1      0      0        2097152
227653       227654        1  1008449.node129      1      0      0        2097152
227654       227655        1  1008449.node129      1      0      0        2097152
227655       227656        1  1008449.node129      1      0      0        2097152
227656       227657        1  1008449.node129      1      0      0        2097152
227657       227658       49  1008449.node129      1      0      0        2097152
227658       227659       49  1008449.node129      1      0      0        2097152
227659       227660       49  1008449.node129      1      0      0        2097152
227660       227661       49  1008449.node129      1      0      0        2097152
227661       227662       49  1008449.node129      1      0      0        2097152
227662       227663       49  1008449.node129      1      0      0        2097152
227663       227664       55  1008449.node129      1      0      0        2097152
227664       227665       55  1008449.node129      1      0      0        2097152
227665       227666       55  1008449.node129      1      0      0        2097152

In [176]:
jobs_to_nodes_whole_data = jobs_to_nodes_whole_data.drop_duplicates(subset=['job_id_string', 'node_id'])

In [177]:
jobs_to_nodes_whole_data.shape[0]

444684

469095

In [178]:
jobs_whole_data = pd.read_csv(infile_jobs)
print("jobs_whole_data contains {} records".format(jobs_whole_data.shape[0]))

jobs_whole_data contains 405771 records


/usr/local/lib/python3.5/dist-packages/IPython/core/interactiveshell.py:3049: DtypeWarning: Columns (11) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [179]:
jobs_whole_data.head(1)

job_id   job_id_string  job_name     queue           start_time       run_start_time             end_time                               user  node_req  cpu_req  mem_req time_req deleted  dependency exit_status  pbs_exit_code
0       1  498458.node129  rth2e6t0  parallel  2014-03-31 00:30:02  2014-03-31 00:30:02  2014-03-31 01:22:40  alupi000@node129.eurora.cineca.it         1       16       14    04:00     NaN         NaN   COMPLETED            NaN

In [180]:
completed_jobs_data = jobs_whole_data[jobs_whole_data.exit_status == 'COMPLETED']
print("completed_jobs_data contains {} records".format(completed_jobs_data.shape[0]))

completed_jobs_data contains 400246 records


how can be? 469095 vs 405771. -> jobs that require more than on node to be executed

In [181]:
jobs_to_nodes_whole_data = jobs_to_nodes_whole_data.drop_duplicates()
print("jobs_to_nodes_whole_data contains {} records".format(jobs_to_nodes_whole_data.shape[0]))

jobs_to_nodes_whole_data contains 444684 records


no duplicates

In [182]:
# join the two dataframe

#merged_jobs_to_nodes = pd.merge(jobs_to_nodes_whole_data, jobs_whole_data, on='job_id_string')
merged_jobs_to_nodes = pd.merge(jobs_to_nodes_whole_data, completed_jobs_data, how='left', on='job_id_string')
print(merged_jobs_to_nodes.shape)

(444684, 22)


In [183]:
merged_jobs_to_nodes.head(2)

job2node_id  node_id   job_id_string  ncpus  ngpus  nmics  mem_requested  job_id  job_name     queue           start_time       run_start_time             end_time                               user  node_req  cpu_req  mem_req time_req deleted  dependency exit_status  pbs_exit_code
0            1       17  498458.node129     16      0      0       14680064     1.0  rth2e6t0  parallel  2014-03-31 00:30:02  2014-03-31 00:30:02  2014-03-31 01:22:40  alupi000@node129.eurora.cineca.it       1.0     16.0     14.0    04:00     NaN         NaN   COMPLETED            NaN
1            2        9  498459.node129     16      0      0       14680064     2.0      CPMD  parallel  2014-03-31 00:36:51  2014-03-31 00:36:51  2014-03-31 00:40:13  framondo@node129.eurora.cineca.it       1.0     16.0     14.0    01:00     NaN         NaN   COMPLETED            NaN

drop where node_id is NaN

In [185]:
df = merged_jobs_to_nodes.groupby('node_id')['job2node_id'].nunique()
print (df)

node_id
1       1875
2        240
3       2388
4       2167
5        484
6       1234
7       2065
8       1761
9        840
10      1677
11      1837
12       420
13      1724
14       376
15       381
16       389
17      2830
18      1315
19      1420
20      1569
21      1462
22         3
23      1480
24      1526
25      1581
26      1149
27      1412
28      1235
29      1006
30       891
31      1018
32       288
33     14814
34     11003
35     13245
36      2340
37      6581
38     10851
39       492
40      7457
41         3
42     10009
44     32866
45     64684
46      9168
47     19797
48      8315
49     16037
50      5360
51     14333
52      1708
53        21
54      1708
55     28978
56     16068
57     14116
58     10630
59     22736
60     10478
61      7847
62      7412
63     12357
64     23153
129       74
Name: job2node_id, dtype: int64


torna l'assenza del node 43 e la scarsità di 22, 41, 53,

In [186]:
regular_data_jobs = merged_jobs_to_nodes[merged_jobs_to_nodes.node_id >= 1]
regular_data_jobs = regular_data_jobs[regular_data_jobs.node_id <= 64]
regular_data_jobs.shape

(444610, 22)

In [187]:
interval_data_jobs = regular_data_jobs[pd.to_datetime(regular_data_jobs['start_time']) > start_time]
interval_data_jobs = interval_data_jobs[pd.to_datetime(interval_data_jobs['end_time']) < end_time]
print("interval_data_jobs contains {} records".format(interval_data_jobs.shape[0]))

interval_data_jobs contains 92937 records


In [188]:
# for cpu in cpus:

cpu = '01'
print(cpu)

01


In [189]:
interval_jobs_node_data = interval_data_jobs[interval_data_jobs['node_id'] == int(cpu)]
print("interval_jobs_node_data contains {} records".format(interval_jobs_node_data.shape[0]))

interval_jobs_node_data contains 480 records


In [190]:
interval_jobs_node_data = interval_jobs_node_data.sort_values('run_start_time')
interval_jobs_node_data.head(5)

job2node_id  node_id   job_id_string  ncpus  ngpus  nmics  mem_requested  job_id         job_name   queue           start_time       run_start_time             end_time                               user  node_req  cpu_req  mem_req time_req deleted  dependency exit_status  pbs_exit_code
2337         2357        1  502459.node129     14      0      0        8388608  2077.0  firestarter.pbs  system  2014-04-01 14:17:20  2014-04-01 14:25:42  2014-04-01 15:25:43  mmorgott@node129.eurora.cineca.it       1.0     14.0      8.0    01:30     NaN         NaN   COMPLETED            NaN
2480         2500        1  502634.node129     16      0      0       13631488  2212.0      lancia.qsub  system  2014-04-01 16:00:23  2014-04-01 17:54:51  2014-04-01 17:54:54  gamati01@node129.eurora.cineca.it       0.0     64.0     52.0    04:00     NaN         NaN   COMPLETED            NaN
2596         2616        1  502645.node129     16      0      0       13631488  2220.0      lancia.qsub  system  2014-04-01 16:07:14  2014-04-01 17:54:55  2014-04-01 21:56:26  gamati01@node129.eurora.cineca.it       8.0    128.0    104.0    04:00     NaN         NaN   COMPLETED            NaN
2605         2625        1  502652.node129      8      0      0        1048576  2224.0         abq_test   debug  2014-04-01 17:08:19  2014-04-01 21:56:27  2014-04-01 21:56:34  amessina@node129.eurora.cineca.it       0.0     16.0      2.0    00:10     NaN         NaN   COMPLETED            NaN
2606         2627        1  502653.node129      8      0      0        1048576  2225.0         abq_test   debug  2014-04-01 17:09:04  2014-04-01 21:57:19  2014-04-01 21:57:19  amessina@node129.eurora.cineca.it       0.0     16.0      2.0    00:10     NaN         NaN   COMPLETED            NaN

In [191]:
infile_node = datadir + "CPUs/" + interval_comment + "/cpu" + cpu + "_" + cpu_suffix + interval_comment +".csv"
node_data = pd.read_csv(infile_node) # Read a comma-separated values (csv) file into DataFrame
print(node_data.shape)

(43200, 6)


In [192]:
node_data['active_cores'] = pd.Series()
node_data['active_jobs'] = pd.Series()
node_data.head(5)

timestamp   01  cpu_id   pow_cpu  pow_dram    pow_pkg  active_cores  active_jobs
0  2014-04-01 00:00:00  1.0     0.0  2.155305  2.507122  17.899565           NaN          NaN
1  2014-04-01 00:01:00  1.0     0.0  2.033380  2.472140  17.762189           NaN          NaN
2  2014-04-01 00:02:00  1.0     0.0  2.037560  2.526941  17.769899           NaN          NaN
3  2014-04-01 00:03:00  1.0     0.0  2.030268  2.487323  17.759689           NaN          NaN
4  2014-04-01 00:04:00  1.0     0.0  1.985083  2.500854  17.683834           NaN          NaN

In [193]:
i = 0
for index, row in node_data.iterrows():
    # print(row)
    running_jobs = interval_jobs_node_data[pd.to_datetime(interval_jobs_node_data['run_start_time']) <= row['timestamp']]
    running_jobs = running_jobs[pd.to_datetime(running_jobs['end_time']) >= row['timestamp']]
    node_data.loc[index,'active_jobs'] = running_jobs.shape[0]
    node_data.loc[index,'active_cores'] = running_jobs['ncpus'].sum()
    if(running_jobs['ncpus'].sum() > 16 and i == 0):
        #print(running_jobs)
        #print(index)
        i = 1
        print(row['timestamp'])
        break
        

    # print(str(running_jobs.shape[0]) + " : " + str(running_jobs['ncpus'].sum()))
    # print(running_jobs.shape[0])
    


In [194]:
running_jobs = interval_jobs_node_data[pd.to_datetime(interval_jobs_node_data['run_start_time']) <= '2014-04-12 15:33:00']
running_jobs[pd.to_datetime(running_jobs['end_time']) >= '2014-04-12 15:33:00']

job2node_id  node_id   job_id_string  ncpus  ngpus  nmics  mem_requested   job_id job_name  queue           start_time       run_start_time             end_time                               user  node_req  cpu_req  mem_req time_req deleted  dependency exit_status  pbs_exit_code
58808        59039        1  586808.node129     16      0      0       14680064  56341.0     test  debug  2014-04-12 15:32:29  2014-04-12 15:32:30  2014-04-12 15:41:26  gbertain@node129.eurora.cineca.it       1.0     16.0     14.0    00:30     DEL         NaN   COMPLETED            NaN

In [195]:
jobs_to_nodes_whole_data[jobs_to_nodes_whole_data['job2node_id'] == 59166]

Empty DataFrame
Columns: [job2node_id, node_id, job_id_string, ncpus, ngpus, nmics, mem_requested]
Index: []

In [196]:
jobs_to_nodes_whole_data[jobs_to_nodes_whole_data['job2node_id'] == 59039]

job2node_id  node_id   job_id_string  ncpus  ngpus  nmics  mem_requested
59038        59039        1  586808.node129     16      0      0       14680064

In [197]:
node_data

timestamp   01  cpu_id    pow_cpu  pow_dram    pow_pkg  active_cores  active_jobs
0      2014-04-01 00:00:00  1.0     0.0   2.155305  2.507122  17.899565           0.0          0.0
1      2014-04-01 00:01:00  1.0     0.0   2.033380  2.472140  17.762189           0.0          0.0
2      2014-04-01 00:02:00  1.0     0.0   2.037560  2.526941  17.769899           0.0          0.0
3      2014-04-01 00:03:00  1.0     0.0   2.030268  2.487323  17.759689           0.0          0.0
4      2014-04-01 00:04:00  1.0     0.0   1.985083  2.500854  17.683834           0.0          0.0
5      2014-04-01 00:05:00  1.0     0.0   2.198175  2.571906  17.991952           0.0          0.0
6      2014-04-01 00:06:00  1.0     0.0   2.145334  2.486568  17.932765           0.0          0.0
7      2014-04-01 00:07:00  1.0     0.0   2.119036  2.432829  17.897167           0.0          0.0
8      2014-04-01 00:08:00  1.0     0.0   2.024416  2.464012  17.742607           0.0          0.0
9      2014-04-01 00:09:00  1.0     0.0   1.994699  2.519761  17.700151           0.0          0.0
10     2014-04-01 00:10:00  1.0     0.0   2.162676  2.496701  17.933494           0.0          0.0
11     2014-04-01 00:11:00  1.0     0.0   2.110549  2.549845  17.880616           0.0          0.0
12     2014-04-01 00:12:00  1.0     0.0   2.098969  2.480298  17.860319           0.0          0.0
13     2014-04-01 00:13:00  1.0     0.0   2.086054  2.493274  17.841488           0.0          0.0
14     2014-04-01 00:14:00  1.0     0.0   2.056178  2.380799  17.790216           0.0          0.0
15     2014-04-01 00:15:00  1.0     0.0   2.037540  2.317000  17.775700           0.0          0.0
16     2014-04-01 00:16:00  1.0     0.0   2.020017  2.324956  17.739720           0.0          0.0
17     2014-04-01 00:17:00  1.0     0.0   2.130488  2.319171  17.901950           0.0          0.0
18     2014-04-01 00:18:00  1.0     0.0   2.095713  2.212518  17.847395           0.0          0.0
19     2014-04-01 00:19:00  1.0     0.0   2.072685  2.204494  17.809104           0.0          0.0
20     2014-04-01 00:20:00  1.0     0.0   2.152255  2.223035  17.921944           0.0          0.0
21     2014-04-01 00:21:00  1.0     0.0   1.950551  2.119778  17.625821           0.0          0.0
22     2014-04-01 00:22:00  1.0     0.0   2.081969  2.096951  17.832327           0.0          0.0
23     2014-04-01 00:23:00  1.0     0.0   2.065678  2.176590  17.808690           0.0          0.0
24     2014-04-01 00:24:00  1.0     0.0   2.040952  2.262151  17.765421           0.0          0.0
25     2014-04-01 00:25:00  1.0     0.0   2.103170  2.358514  17.870106           0.0          0.0
26     2014-04-01 00:26:00  1.0     0.0   2.060655  2.217768  17.795701           0.0          0.0
27     2014-04-01 00:27:00  1.0     0.0   2.071491  2.261612  17.813526           0.0          0.0
28     2014-04-01 00:28:00  1.0     0.0   2.062681  2.284717  17.804446           0.0          0.0
29     2014-04-01 00:29:00  1.0     0.0   2.069265  2.271264  17.812877           0.0          0.0
30     2014-04-01 00:30:00  1.0     0.0   2.138813  2.332790  17.913306           0.0          0.0
31     2014-04-01 00:31:00  1.0     0.0   2.062012  2.207166  17.798295           0.0          0.0
32     2014-04-01 00:32:00  1.0     0.0   2.070067  2.245513  17.817314           0.0          0.0
33     2014-04-01 00:33:00  1.0     0.0   2.057159  2.207162  17.800431           0.0          0.0
34     2014-04-01 00:34:00  1.0     0.0   2.054704  2.198253  17.792470           0.0          0.0
35     2014-04-01 00:35:00  1.0     0.0   2.055837  2.178795  17.795097           0.0          0.0
36     2014-04-01 00:36:00  1.0     0.0   2.013259  2.155583  17.728471           0.0          0.0
37     2014-04-01 00:37:00  1.0     0.0   1.994091  2.076252  17.696429           0.0          0.0
38     2014-04-01 00:38:00  1.0     0.0   2.048104  2.119855  17.788879           0.0          0.0
39     2014-04-01 00:39:00  1.0     0.0   2.032030  2.279

In [198]:
node_data['active_jobs'].max()

7.0

In [199]:
jobs_to_nodes_whole_data['ncpus'].max()

16

In [200]:
df = node_data.groupby('active_cores')['timestamp'].nunique()
print(df)

active_cores
0.0     38478
1.0       639
2.0       162
3.0         7
4.0       103
5.0         1
6.0         1
7.0         6
8.0       162
12.0      313
14.0       65
16.0     3263
Name: timestamp, dtype: int64


In [155]:
node_data.iloc[[16773]]

timestamp   01  cpu_id    pow_cpu  pow_dram    pow_pkg  active_cores  active_jobs
16773  2014-04-12 15:33:00  1.0     0.0  40.137644  4.480388  58.031166          32.0          2.0